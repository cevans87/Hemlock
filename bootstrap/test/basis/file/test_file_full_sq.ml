open Basis.Rudiments
open Basis

external nop_submit: unit -> u64 = "hm_basis_file_nop_submit_inner"
external user_data_decref: u64 -> unit = "hm_basis_file_user_data_decref"

external user_data_pp: File.Open.t -> unit = "hm_basis_file_user_data_pp"
external sqring_pp: unit -> unit = "hm_basis_file_sqring_pp"
external ioring_pp: unit -> unit = "hm_basis_file_ioring_pp"

let () =
  let rec submit_nops i n = begin
    match i < n with
    | false -> ()
    | true -> let _ = user_data_decref (nop_submit ()) in
      submit_nops (i + 1L) n
  end in
  submit_nops 0L 32L;
  sqring_pp ();
  let s = File.Open.submit ~flag:File.Flag.RW (Bytes.Slice.of_string_slice
      (String.C.Slice.of_string "file")) in
  user_data_pp s;
  ioring_pp ();
  let _ = File.Open.complete_hlt s in
  user_data_pp s;
  sqring_pp ();
  ()