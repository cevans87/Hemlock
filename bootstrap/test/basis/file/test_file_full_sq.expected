sqring:
  head:         0
  tail:         32
  ring_entries: 32
  ring_mask:    31
  flags:        0
  array: [0 ... 31]
  sqes:
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
    sqe:
      flags:    0
      ioprio:   0
      fd:       0
      opcode:   IORING_OP_NOP
      user_data:
        refcount: 1
        opcode:   IORING_OP_NOP
        cqe:
          res:       0
          flags:     0
          user_data: NULL
user_data:
  refcount: 2
  pathname: "file"
  opcode:   IORING_OP_OPENAT
  cqe:
    res:       0
    flags:     0
    user_data: NULL
ioring:
  fd:           3
  cqring:
    head:         32
    tail:         32
    ring_entries: 64
    ring_mask:    63
    cqes:
  sqring:
    head:         32
    tail:         33
    ring_entries: 32
    ring_mask:    31
    flags:        0
    array: [0 ... 31]
    sqes:
      sqe:
        flags:    0
        ioprio:   0
        fd:       -100
        opcode:   IORING_OP_OPENAT
        user_data:
          refcount: 2
          pathname: "file"
          opcode:   IORING_OP_OPENAT
          cqe:
            res:       0
            flags:     0
            user_data: NULL
user_data:
  refcount: 1
  pathname: NULL
  opcode:   IORING_OP_OPENAT
  cqe:
    res:       4
    flags:     0
    user_data: <same as parent>
sqring:
  head:         33
  tail:         33
  ring_entries: 32
  ring_mask:    31
  flags:        0
  array: [0 ... 31]
  sqes:
